Terminal.Gui/Views/GraphView.cs

```diff
@@ -0,0 +1,1913 @@
+				yield break;
+			}
+
+			int labels = 0;
+
+			int x = GetAxisXPosition (graph, bounds);
+
+			for (int i = 0; i < bounds.Height; i++) {
+
+				// what bit of the graph is supposed to go here?
+				var graphSpace = graph.ScreenToGraphSpace (x, i);
+
+				if (Math.Abs (graphSpace.Y) % Increment < graph.CellSize.Y) {
+					var toRender = new AxisIncrementToRender (Orientation, new Point (x, i), graphSpace);
+
+					// and the label (if we are due one)
+					if (ShowLabelsEvery != 0) {
+
+						// if this increment also needs a label
+						if (labels++ % ShowLabelsEvery == 0) {
+							toRender.Text = LabelGetter (toRender);
+						};
+
+						yield return toRender;
+					}
+
+				}
+
+			}
+		}
+
+		/// <summary>
+		/// Returns the X screen position of the origin (typically 0,0) of graph space.
+		/// Return value is bounded by the screen i.e. the axis is always rendered even
+		/// if the origin is offscreen.
+		/// </summary>
+		/// <param name="graph"></param>
+		/// <param name="bounds"></param>
+		private int GetAxisXPosition (GraphView graph, Rect bounds)
+		{
+			// find the origin of the graph in screen space (this allows for 'crosshair' style
+			// graphs where positive and negative numbers visible
+			var origin = graph.GraphSpaceToScreen (new PointD (0, 0));
+
+			// float the Y axis so that it accurately represents the origin of the graph
+			// but anchor it to left/right if the origin is offscreen
+			return Math.Min (Math.Max ((int)graph.MarginLeft, origin.X), bounds.Width - 1);
+		}
+	}
+
+
+	/// <summary>
+	/// A location on an axis of a <see cref="GraphView"/> that may
+	/// or may not have a label associated with it
+	/// </summary>
+	public class AxisIncrementToRender {
+
+		/// <summary>
+		/// Direction of the parent axis
+		/// </summary>
+		public Orientation Orientation { get; }
+
+		/// <summary>
+		/// Location in the <see cref="Axis"/> that the axis increment appears
+		/// </summary>
+		public Point ScreenLocation { get; }
+
+		/// <summary>
+		/// The volume of graph that is represented by this screen coordinate
+		/// </summary>
+		public RectangleD GraphSpace { get; }
+
+		private string _text = "";
+
+		/// <summary>
+		/// The text (if any) that should be displayed at this axis increment
+		/// </summary>
+		/// <value></value>
+		public string Text {
+			get => _text;
+			internal set { _text = value ?? ""; }
+		}
+
+		/// <summary>
+		/// Describe a new section of an axis that requires an axis increment
+		/// symbol and/or label
+		/// </summary>
+		/// <param name="orientation"></param>
+		/// <param name="screen"></param>
+		/// <param name="graphSpace"></param>
+		public AxisIncrementToRender (Orientation orientation, Point screen, RectangleD graphSpace)
+		{
+			Orientation = orientation;
+			ScreenLocation = screen;
+			GraphSpace = graphSpace;
+		}
+	}
+
+	/// <summary>
+	/// Rectangle class based on the exact floating point primative 'decimal'
+	/// </summary>
+	public class RectangleD {
```

Comment:
Did you code these floating point graphic primiatives yourself or copy the code from e.g. `System.Drawing's RectangleF`?

We should avoid re-implementing something that others have done before (and tested). So I'm hoping you say you copied the code. e.g. from https://docs.microsoft.com/en-us/dotnet/api/system.drawing.rectanglef?view=net-5.0

Why `decimal` v. `float`?

As you have implemented this, these classes are at the top `Terminal.Gui` class hierarchy. This means we are explicitly documenting them such that they can be used by others OUTSIDE of the use of `GraphView`. I think we need to very carefully consider whether this is a good idea or not.

If it IS a good idea then the bar for exposing them as `public` is MUCH higher and we should have superb unit tests for every aspect of the API. 

If it is NOT a good idea then we should consider how we would either NOT make them public or ensure they are contained with a namespace that identifies them as part of `GraphView`.

Reply:
Yes the code for RectangleD is just RectangleF with a find/replace on `float` for `double` http://www.dotnetframework.org/default.aspx/DotNET/DotNET/8@0/untmp/whidbey/REDBITS/ndp/fx/src/CommonUI/System/Drawing/Advanced/RectangleF@cs/1/RectangleF@cs 

Since dotnet is open source already I didn't see a problem with copying it.  I've put a comment in the main thread about why I switched from `float` to `decimal` and am open to exploring that choice.